rfcs/text/0000-metadata-integration-test.md

```diff
@@ -0,0 +1,106 @@
+
+## Detailed Design
+
+Below we describe a design that will transition DCP from the current "waterfall" software/metadata integration model to a more observable and interactive integration model, informed by integration tests that display the true compatibility state of the system.
+
+Setting this plan will also better define the interaction between the metadata team and the rest of the DCP. Processes and responsibilities for resolving any integration issues that arise will be clear.
+
+### Design Alternative #1
+
+Software development proceeds in the dev branch of each component independently. Changes are merged into integration branch of each component only through PRs which cause system-wide integration tests to run. PRs are merged automatically if tests pass (but can be reverted if downstream breakage occurs). PRs may not be merged if tests do not pass.
+
+![Option 1](../images/0000-metadata-integration-test-opt1.png)
+
+*Here DSS Dev is shown as an example of other components following the same process.*
+
+Data with the new schema version should not be uploaded into the `integration` or subsequent deployment stages until the test passes and the `master` branch is promoted to `integration`.
+
+#### Pros
+
+* "Test then merge"
+* PR status will not require wranglers to know about downstream systems
+* Small non-breaking changes can be merged quickly (out-of-order) and with confidence
+* Greatly reduces probability of deployment stages past integration breaking
+* More certainty for developers on behavior of the system
+* Issues caused by metadata changes are surfaced early
+* All DCP projects are integrated in the same way
+
+#### Cons
+
+* May be difficult to revert test versions of components and metadata (not all infra deployments are easily reversible)
+* Requires "serializing" the dcp-wide integration test so that one PR can run after another
+
+### Design Alternative #2 
+
+Software development proceeds in the dev branch of each component independently.
+When a metadata schema version has passed unit tests in `dev` and its committers wish to promote it, they make a PR against their integration branch.
+The PR triggers a dcp-wide test in the integration environment.
+Metadata schema changes can be merged from `dev` to `integration` only if the integration test suite passes.
+
+![Option 2](../images/0000-metadata-integration-test-opt2.png)
+
+Data with the new schema version should not be uploaded into the `integration` or subsequent deployment stages until the test passes and the `master` branch is promoted to `integration`.
+
+#### Pros
+
+* "Test then merge"
+* PR status will not require wranglers to know about downstream systems
+* Small non-breaking changes can be merged quickly and with confidence
+* Greatly reduces probability of deployment stages past `dev` breaking
+
+#### Cons
+
+* May introduce another deployment stage
+* Treats metadata schema integration differently from other DCP projects
```

Comment:
Why is it a Con that the metadata schemas are treated differently than other DCP components?

Reply:
It's not necessarily a con. I don't see it as such. It just felt important to acknowledge the design as asymmetrical. It is.